[breaking][libra framework] add event on account creation

sblackshear opened the pull request:
- New `CreateAccountEvent` struct that records the created address
- Had to refactor LibraAccount::initialize() a bit to ensure that we can emit an event for the creation of LibraRoot
- Had to refactor the event verification part of VM genesis since the number of events now depends on the # of created accounts
- Added tests for each of the create_*_account functions
- Added Rust wrapper for CreateAccountEvent

tzakian commented on language/tools/vm-genesis/src/lib.rs:
```diff
@@ -510,29 +514,27 @@ fn reconfigure(session: &mut Session<StateViewCache>) {
+        .iter()
+        .find(|e| e.key() == &NewEpochEvent::event_key())
+        .expect("Couldn't find NewEpochEvent");
+    assert_eq!(*new_epoch_event.key(), NewEpochEvent::event_key());
```
What would you think of adding an additional check that there's only one new epoch event?

sblackshear replied:
Very good idea, would not want to make that mistake (and it'd be easy to make!).

sblackshear commented on language/stdlib/modules/LibraAccount.move:
```diff
@@ -186,40 +193,18 @@ module LibraAccount {
             lr_account,
             copy dummy_auth_key_prefix,
         );
-
-        assert(
-            !exists<AccountOperationsCapability>(CoreAddresses::LIBRA_ROOT_ADDRESS()),
-            Errors::already_published(EACCOUNT_OPERATIONS_CAPABILITY)
-        );
-        move_to(
-            lr_account,
-            AccountOperationsCapability {
-                limits_cap: AccountLimits::grant_mutation_capability(lr_account),
-            }
-        );
-
-        assert(
-            !exists<LibraWriteSetManager>(CoreAddresses::LIBRA_ROOT_ADDRESS()),
-            Errors::already_published(EWRITESET_MANAGER)
-        );
-        move_to(
-            lr_account,
-            LibraWriteSetManager {
-                upgrade_events: Event::new_event_handle<Self::UpgradeEvent>(lr_account),
-            }
-        );
```
Unfortunately, this won't wok because emitting the event (which happens in `create_libra_root_account`) requires `<THIS_CODE>` to be run first, specifically, the `AccountOperationsCapability` part.

I actually felt good about moving this code into `create_libra_root_account` because the pattern we use elsewhere is to publish X role-specific resources in `create_X_account`. But certainly open to alternatives!

tzakian replied:
ahhh I see! For some reason I didn't see that the event emission is happening in `make_account`. 

I'm fine with this as-is since I can't think of any other alternatives that wouldn't make the code more complicated.

sblackshear commented on language/stdlib/modules/LibraAccount.move:
```diff
@@ -123,6 +124,12 @@ module LibraAccount {
+    /// Message for creation of a new account
+    struct CreateAccountEvent {
+        /// Address of the account that was created
+        created: address,
```
I love it except that role id's will no longer be opaque. But I'm ok with that because they can already be read off-chain...

tzakian reviewed (approved):
LGTM!

sblackshear commented:
/canary

bors-libra commented:
:broken_heart: Test Failed - [Build images and run cluster test](https://github.com/libra/libra/runs/1161029165)

sblackshear commented:
/canary

bors-libra commented:
:broken_heart: Test Failed - [Build images and run cluster test](https://github.com/libra/libra/runs/1161794947)

sblackshear commented:
/canary

bors-libra commented:
:broken_heart: Test Failed - [Build images and run cluster test](https://github.com/libra/libra/runs/1162556724)

sblackshear commented:
@andll: should I be concerned by `/canary` continuing to fail here?

andll commented:
Yes, it seems that this PR does not work in cluster test - validators do not start up
Did you try to examine validator logs to see what is going on there?

sblackshear commented:
> Yes, it seems that this PR does not work in cluster test - validators do not start up
> Did you try to examine validator logs(see the link in cluster test bot comment) to see what is going on there?

Good call! It's failing here: 

```
execution/executor/src/db_bootstrapper.rs:        .ok_or_else(|| format_err!("Genesis transaction must emit a epoch change."))?;
```

, but I'm puzzled why. The event handle changed (which could cause the executor to fail to recognize that the event handle for NewEpochEvent changed), but I updated the hardcoded handle ID and there are several other tests that would fail if this was not done correctly (as I learned the hard way...). I also took a look at https://github.com/libra/libra/pull/6042, which also changed this handle to make sure this PR is updating all of the same code.

Is it possible that `/canary` runs new code against old state (or vice versa) in a way that might trigger the error?

andll commented:
Yea, you are right. Canary does same test as landing would do. In this case it’s compatibility test. Need to add breaking label to commit.

Sent from my iPhone

On Sep 24, 2020, at 7:00 PM, Sam Blackshear <notifications@github.com> wrote:

﻿

Yes, it seems that this PR does not work in cluster test - validators do not start up
Did you try to examine validator logs(see the link in cluster test bot comment) to see what is going on there?

Good call! It's failing here:

execution/executor/src/db_bootstrapper.rs:        .ok_or_else(|| format_err!("Genesis transaction must emit a epoch change."))?;


, but I'm puzzled why. The event handle changed (which could cause the executor to fail to recognize that the event handle for NewEpochEvent changed), but I updated the hardcoded handle ID and there are several other tests that would fail if this was not done correctly (as I learned the hard way...). I also took a look at #6042<https://github.com/libra/libra/pull/6042>, which also changed this handle to make sure this PR is updating all of the same code.

Is it possible that /canary runs new code against old state (or vice versa) in a way that might trigger the error?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/libra/libra/pull/6156#issuecomment-698678075>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AAWVVQMFAMSRASRRBB7BNNTSHP2UFANCNFSM4RWSVAZQ>.

sblackshear commented:
Ahh I forgot about that. Thanks!

sblackshear commented:
/canary

bors-libra commented:
:sunny: Canary successful

sblackshear commented:
/land

bors-libra commented:
:lock: Merge Conflict

sblackshear commented:
/land

sblackshear commented:
/cancel

sblackshear commented:
/land
